YAML config file parsing behaviour changed between 5.x and master

The root of this change is https://github.com/elastic/elasticsearch/pull/22225

The following configuration file works in 5.x (tested with a 5.4 snapshot from https://oss.sonatype.org/content/repositories/snapshots/org/elasticsearch/distribution/zip/elasticsearch/5.4.0-SNAPSHOT/), but throws an exception in master, tested with a snapshot from https://oss.sonatype.org/content/repositories/snapshots/org/elasticsearch/distribution/zip/elasticsearch/6.0.0-alpha1-SNAPSHOT/

```YAML
node:
  name: node1

node:
  attr:
    rack: r1
```

This stacktrace is shown on master


```
Exception in thread "main" SettingsException[Failed to load settings from /tmp/test-nesting/elasticsearch-6.0.0-alpha1-SNAPSHOT/config/elasticsearch.yml]; nested: JsonParseException[Duplicate field 'node'
...
; line: 4, column: 5]];
	at org.elasticsearch.node.InternalSettingsPreparer.prepareEnvironment(InternalSettingsPreparer.java:104)
	at org.elasticsearch.cli.EnvironmentAwareCommand.createEnv(EnvironmentAwareCommand.java:63)
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:58)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:122)
	at org.elasticsearch.cli.Command.main(Command.java:88)
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:91)
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:84)
Caused by: com.fasterxml.jackson.core.JsonParseException: Duplicate field 'node'
 at [Source: node:
  name: node1

node:
  attr:
    rack: r1
...
; line: 4, column: 5]
	at com.fasterxml.jackson.core.json.JsonReadContext._checkDup(JsonReadContext.java:204)
	at com.fasterxml.jackson.core.json.JsonReadContext.setCurrentName(JsonReadContext.java:198)
	at com.fasterxml.jackson.dataformat.yaml.YAMLParser.nextToken(YAMLParser.java:372)
	at org.elasticsearch.common.xcontent.json.JsonXContentParser.nextToken(JsonXContentParser.java:52)
	at org.elasticsearch.common.settings.loader.XContentSettingsLoader.serializeObject(XContentSettingsLoader.java:105)
	at org.elasticsearch.common.settings.loader.XContentSettingsLoader.load(XContentSettingsLoader.java:75)
	at org.elasticsearch.common.settings.loader.XContentSettingsLoader.load(XContentSettingsLoader.java:52)
	at org.elasticsearch.common.settings.loader.YamlSettingsLoader.load(YamlSettingsLoader.java:50)
	at org.elasticsearch.common.settings.Settings$Builder.loadFromStream(Settings.java:1046)
	at org.elasticsearch.common.settings.Settings$Builder.loadFromPath(Settings.java:1035)
	at org.elasticsearch.node.InternalSettingsPreparer.prepareEnvironment(InternalSettingsPreparer.java:102)
	... 6 more
```

I do consider this a good thing, just something we should be aware of that needs to be kept in mind when upgrading (part of the upgrade docs, as it is hard to warn about before the upgrade I suppose).

Feel free to close without comment if not considered an issue.

--- a/docs/reference/migration/migrate_6_0/settings.asciidoc
+++ b/docs/reference/migration/migrate_6_0/settings.asciidoc
@@ -1,6 +1,32 @@
 [[breaking_60_settings_changes]]
 === Settings changes
 
+==== Duplicate keys in configuration file
+
+In previous versions of Elasticsearch, the configuration file was allowed to
+contain duplicate keys. For example:
+
+[source,yaml]
+--------------------------------------------------
+node:
+  name: my-node
+
+node
+  attr:
+    rack: my-rack
+--------------------------------------------------
+
+In Elasticsearch 6.0.0, this is no longer permitted. Instead, this must be
+specified in a single key as:
+
+[source,yaml]
+--------------------------------------------------
+node:
+  name: my-node
+  attr:
+    rack: my-rack
+--------------------------------------------------
+
 ==== Coercion of boolean settings
 
 Previously, Elasticsearch recognized the strings `true`, `false`, `on`, `off`, `yes`, `no`, `0`, `1` as booleans. Elasticsearch 6.0

